### DISCUSSION TITLE
How to hide `mouse cursor` while doing OCR using NormCap ?

### DISCUSSION THREAD

**Discussion by KAGEYAM4** [dynobo/normcap#459] [Q&A] [answered]
i am on ArchLinux KDE Plasma, when i trigger `normcap` either through `shortcut` or `comman-line`, the `cursor` is part of OCR selection area, is there a way to hide it ?

**Comment by dynobo** [comment]
Hi @KAGEYAM4, I'm not sure if I understand correctly: Do you think the cursor interferes with the OCR functionality, e.g. is the detection outcome unexpected? Or is it a usability issue and the cursor just makes it more difficult for you to select the area?

It _should_ look like below: the cursor should be cross-hair in the corner, and it shouldn't really cover the selection area:
![Screenshot from 2023-06-04 16-43-36](https://github.com/dynobo/normcap/assets/11071876/9326c94a-b417-42bc-aad9-3291bf08ab75)

Does it look different for you? If yes, could you post a screenshot? 
(On many systems this can be done via pressing the `<print>`- or `<print screen>`-key. If that doesn't work for you, you could try to trigger a delayed screenshot.

**Reply by KAGEYAM4** [reply]
Sorry for Late Reply, This is how it looks on my side when `normcap` is triggered, `a bouncing loading animation shows up beside my cursor and that also becomes part of the screenshot where OCR has to be performed resulting in text getting hidden`. Here `man, info & apropos`, the letter `n` got hidden.

![image](https://github.com/dynobo/normcap/assets/75798544/2662974a-20eb-4c39-9466-a64e50ad3e39)

**Reply by dynobo** [reply]
@KAGEYAM4 , Ah, now I see! Thanks for additional screenshot! 

That's interesting! I'm also using Arch Linux, but with Gnome. For me, the screenshot does not include the cursor. 
May I ask if you are using "Xorg" or "Wayland" for your sessions? ( `echo $XDG_SESSION_TYPE` should help you to find that out)

And could you please try to launch NormCap with the debug flag `normcap -v debug` from the terminal [to print debug output](https://dynobo.github.io/normcap/#faqs-debug)? Watch out for a line in the log which says **"Select capture method {QT | DBUS portal | DBUS shell}**. Which of the three methods does NormCap chose on your system? Depending on the method I might be able to somehow actively hide the cursor in the screenshot....

**Comment by KAGEYAM4** [comment]
```
$ echo $XDG_SESSION_TYPE
wayland
```

DBUS portal - `08:18:39 - DEBUG   - normcap.screengrab:37 - Select capture method DBUS portal`

<details>
<summary>normcap -v debug output</summary>

```
$ normcap -v debug
08:18:39 - INFO    - normcap:29 - Start NormCap v0.4.2
08:18:39 - DEBUG   - normcap:84 - Set QT_QPA_PLATFORM=wayland
08:18:39 - DEBUG   - normcap.gui.tray:60 - System info:
{'cli_args': '/usr/bin/normcap -v debug', 'is_briefcase_package': False, 'is_flatpak_package': False, 'platform': 'linux', 'pyside6_version': '6.5.1.1', 'qt_version': '6.5.1', 'qt_library_path': '/usr/lib/qt6/plugins, /usr/bin', 'config_directory': PosixPath('/home/phoenix/.config/normcap'), 'normcap_version': '0.4.2', 'ressources_path': PosixPath('/usr/lib/python3.11/site-packages/normcap/resources'), 'tesseract_path': PosixPath('/usr/bin/tesseract'), 'tessdata_path': None, 'envs': {'TESSDATA_PREFIX': None, 'LD_LIBRARY_PATH': None}, 'desktop_environment': <DesktopEnvironment.KDE: 2>, 'display_manager_is_wayland': True, 'screens': [Screen(is_primary=True, device_pixel_ratio=1.0, rect=Rect(left=0, top=0, right=1366, bottom=768), index=0, screenshot=None)]}
08:18:39 - DEBUG   - normcap.gui.tray:334 - Listen on local socket v0.4.2-normcap.
08:18:39 - DEBUG   - normcap.gui.settings:128 - Skip update of non existing setting (cli_mode: False)
08:18:39 - DEBUG   - normcap.screengrab:37 - Select capture method DBUS portal
08:18:39 - DEBUG   - normcap.screengrab.dbus_portal:194 - Request screenshot with interactive=False
08:18:39 - DEBUG   - normcap:211 - [QT] qtwarningmsg - could not connect "org.freedesktop.portal.request" to "on_exception"
08:18:39 - DEBUG   - normcap.screengrab.dbus_portal:77 - Request accepted
08:18:39 - DEBUG   - normcap.screengrab.dbus_portal:104 - Parse response
08:18:39 - DEBUG   - normcap.screengrab.utils:26 - Virtual geometry width: 1366
08:18:39 - DEBUG   - normcap.screengrab.utils:27 - Image width: 1366
08:18:39 - DEBUG   - normcap.screengrab.utils:28 - Resize ratio: 1.0
08:18:39 - DEBUG   - normcap.gui.utils:22 - Save debug image as /tmp/normcap/1686365319.79354_raw_screen0.png
08:18:39 - DEBUG   - normcap.gui.window:131 - Create window for screen 0
08:18:39 - DEBUG   - normcap.gui.window:193 - Set window of screen 0 to fullscreen
08:18:39 - DEBUG   - normcap:211 - [QT] qtwarningmsg - qsystemtrayicon::setvisible: no icon set
08:18:40 - DEBUG   - normcap.gui.window:184 - Move window 0 to (left=0, top=0, right=1366, bottom=768)
08:18:40 - DEBUG   - normcap.ocr.tesseract:23 - Tesseract command output:
List of available languages in "/usr/share/tessdata/" (2):
afr
osd
08:19:13 - DEBUG   - normcap.gui.tray:291 - Hide 1 window
08:19:13 - INFO    - normcap.gui.tray:194 - Crop image to region (159, 243, 258, 266)
08:19:13 - DEBUG   - normcap.gui.utils:22 - Save debug image as /tmp/normcap/1686365353.1062262_cropped.png
08:19:13 - DEBUG   - normcap.gui.tray:219 - Start OCR
08:19:13 - DEBUG   - normcap.ocr.enhance:76 - Scale image x3.2
08:19:13 - DEBUG   - normcap.ocr.enhance:54 - Pad image by 80px
08:19:13 - DEBUG   - normcap.ocr.enhance:92 - Invert image
08:19:13 - DEBUG   - normcap.ocr.recognize:35 - Run Tesseract on image of size (474, 233) with args:
TessArgs(tessdata_path=None, lang='afr', oem=<OEM.DEFAULT: 3>, psm=<PSM.AUTO_OSD: 1>)
08:19:13 - DEBUG   - normcap.ocr.tesseract:23 - Tesseract command output:

08:19:13 - DEBUG   - normcap.ocr.recognize:44 - OCR result:
OcrResult(tess_args=TessArgs(tessdata_path=None, lang='afr', oem=<OEM.DEFAULT: 3>, psm=<PSM.AUTO_OSD: 1>), words=[{'level': 5, 'page_num': 1, 'block_num': 1, 'par_num': 1, 'line_num': 1, 'word_num': 1, 'left': 83, 'top': 84, 'width': 105, 'height': 51, 'conf': 21.830986, 'text': 'rMark'}, {'level': 5, 'page_num': 1, 'block_num': 1, 'par_num': 1, 'line_num': 1, 'word_num': 2, 'left': 197, 'top': 111, 'width': 40, 'height': 24, 'conf': 96.706146, 'text': 'as'}, {'level': 5, 'page_num': 1, 'block_num': 1, 'par_num': 1, 'line_num': 1, 'word_num': 3, 'left': 247, 'top': 84, 'width': 144, 'height': 51, 'conf': 26.138306, 'text': 'answer—`'}], image=<PySide6.QtGui.QImage(QSize(474, 233),format=QImage::Format_RGB32,depth=32,devicePixelRatio=1,bytesPerLine=1896,sizeInBytes=441768) at 0x7f8fee846e00>, magic_scores={}, parsed='')
08:19:13 - INFO    - normcap.ocr.magics.email_magic:33 - 0 emails found 
08:19:13 - DEBUG   - normcap.ocr.magics.email_magic:41 - 0/15 (0.0) chars in emails
08:19:13 - INFO    - normcap.ocr.magics.url_magic:55 - 0 URLs found 
08:19:13 - DEBUG   - normcap.ocr.magics.url_magic:63 - 0/17 (0.0) chars in urls
08:19:13 - DEBUG   - normcap.ocr.magics.magic:70 - Magic scores:
{'SingleLineMagic': 50, 'MultiLineMagic': 0, 'ParagraphMagic': 0.0, 'EmailMagic': 0.0, 'UrlMagic': 0.0}
08:19:13 - DEBUG   - normcap.ocr.recognize:48 - Parsed text:
rMark as answer—`
08:19:13 - DEBUG   - normcap.gui.utils:22 - Save debug image as /tmp/normcap/1686365353.9272773_enhanced.png
08:19:13 - INFO    - normcap.gui.tray:237 - Text from OCR:
rMark as answer—`
08:19:13 - DEBUG   - normcap.clipboard.linux:38 - Select clipboard method wl-copy
08:19:13 - DEBUG   - normcap.gui.tray:267 - Copy text to clipboard
08:19:13 - DEBUG   - normcap.gui.notifier:85 - Send notification via notify-send
08:19:18 - INFO    - normcap.gui.tray:499 - Exit normcap (notification sent delaying exit)
08:19:18 - DEBUG   - normcap.gui.tray:500 - Debug images saved in /tmp/normcap
```
</details>

**Reply by dynobo** [reply]
Everything actually looks like it should, and the log is the same as on my Arch... 

I wonder: if I remember correctly, the Gnome default "Screenshot" tool has check box to include/exclude the pointer from the screenshot. Could you please check if there is something similar on KDE, and at which state it is? (Theoretically, that setting in "Screenshot" should not apply for other apps, but given its deep integration in Gnome, I also wouldn't be surprised, if it does...)

**Reply by dynobo** [reply, marked as answer]
Hi @KAGEYAM4 ,

I just double checked: the method that can be used to take screenshots on Wayland [does _not_ contain an options to hide the mouse pointer](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.Screenshot). 

I also tried on my system, if the setting in Gnome's "Screenshot" tool makes a difference, but for me it does not.

So I'm kind of running out of ideas here. My best guess is, that KDE manipulates the mouse pointer in a weired way to be able to show this animation, and that the animation is somehow technically not part of the mouse pointer.

In the meanwhile: is moving the mouse pointer out of your way _before_ starting NormCap is viable workaround for you?

**Reply by KAGEYAM4** [reply]
Thank you for looking into this.

In spectacle the option for `include mouse pointer` is unchecked.

No worries, i will use this `moving the mouse pointer out of your way before starting NormCap`.

**Reply by dynobo** [reply]
Okay, I think I found the reason: KDE has something called "Launch Feedback". It seems to control this animation, and it seems to be time depend: It will be visible _x_ seconds, independently of how long the application actually take to be responsive.

So, nothing I can do from NormCap side, but you might configure this feature according to your needs. It seems to be well hidden in the settings:

![launcher settings](https://forum.endeavouros.com/uploads/default/original/3X/c/d/cd3de865afbad1b7b1fb03a3d448a8f2973543e8.png)
